File: 백준/Silver/1446. 지름길/지름길.py
import sys

input = sys.stdin.readline

n,d = map(int , input().split(' '))

roads = sorted([list(map(int,input().split(' '))) for i in range(1,1+n)],key=lambda x:x[1])
dp = [i for i in range(0,d+1)]

for [s,e,v] in list(filter(lambda x:x[1]<=d, roads)):
  dp[e] = min(dp[e],dp[s]+v)
  for i in range(e+1,d+1):
    dp[i] = min(dp[i],dp[e]+i-e)

print(dp[d])

File: 백준/Silver/1446. 지름길/지름길.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [N, D] = input[0].split(" ").map(Number)
const roads = input
  .slice(1)
  .map((row) => row.split(" ").map(Number))
  .filter(([_, end, v]) => end <= D)
  .sort((a, b) => a[1] - b[1])

const dp = Array.from({ length: D + 1 }, (v, index) => index)

for (let [s, e, value] of roads) {
  dp[e] = Math.min(dp[e], dp[s] + value)

  for (let i = e + 1; i <= D; i++) {
    if (dp[i] > dp[e] + (i - e)) {
      dp[i] = dp[e] + (i - e)
    }
  }
}

console.log(dp[D])
